[Pal/Linux-SGX] Detect invalid manifest.sgx earlier

jiazhang0 opened the pull request:
[Pal/Linux-SGX] Refactor main function of pal_loader
    
    - Avoid opening manifest file twice
      If the specified file is a manifest, its fd can be sent to load_enclave()
      directly.
    
    - Kill hard-coding constant
      Use static_strlen(foo) instead.
    
    - Add error handling
      Check the return value of alloc_concat(), and use strendswith() to prevent
      from buffer underflow caused by short base name of input file path.
    
    - Enhance readability
      Include renaming variables and comment changes

Signed-off-by: Jia Zhang <zhang.jia@linux.alibaba.com>

<!-- Please fill in the following form before submitting this PR and ensure that your code follows our [coding style guideline](../blob/master/CODESTYLE.md). -->

## Affected components

- [ ] README and global configuration
- [x] Linux PAL
- [ ] SGX PAL
- [ ] FreeBSD PAL
- [ ] Common PAL code
- [ ] Library OS (i.e., SHIM), including GLIBC

## Description of the changes <!-- (reasons and measures) -->
Bug fix

## How to test this PR? <!-- (if applicable) -->
```
# cd LibOS/shim/test/native
# echo foofoofoo > test
# ./pal_loader SGX test
Cannot open manifest test.manifest.sgx
```

With the patch applied:
```
Cannot open manifest file: test.manifest.sgx
USAGE: /root/graphene/Runtime/pal-Linux-SGX [executable|manifest] args ...
```

bash regression also goes pass.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/oscarlab/graphene/1007)
<!-- Reviewable:end -->

dimakuv commented:
Jenkins, test this please

jiazhang0 commented:
@dimakuv @yamahata Revised as suggested.

jiazhang0 commented:
> > _[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1007)_ status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @jiazhang0)
> 
> _[a discussion](https://reviewable.io/reviews/oscarlab/graphene/1007#-LpWmwq19BA6ydTMBmTl:-LpWmwq19BA6ydTMBmTm:bcr2xga) (no related file):_
> Thanks for refactoring! One small ask: could you please use 100-characters line widths? There is no need for line wraps in this code...

Updated.

dimakuv commented:
@jiazhang0 Letting you know that #971 was merged today. You can rebase and proceed with your fix.

jiazhang0 commented:
> @jiazhang0 Letting you know that #971 was merged today. You can rebase and proceed with your fix.

A quick diff is given:
```
diff --git a/Pal/src/host/Linux-SGX/sgx_main.c b/Pal/src/host/Linux-SGX/sgx_main.c
index 77f10d71..4d7885d8 100644
--- a/Pal/src/host/Linux-SGX/sgx_main.c
+++ b/Pal/src/host/Linux-SGX/sgx_main.c
@@ -1020,9 +1020,9 @@ int main (int argc, char ** argv, char ** envp)
          * removed.
          */
         size_t exec_len = strlen(exec_uri);
-        if (strcmp_static(exec_uri + exec_len - strlen(".manifest"), ".manifest")) {
+        if (strendswith(exec_uri, ".manifest")) {
             exec_uri[exec_len - strlen(".manifest")] = '\0';
-        } else if (strcmp_static(exec_uri + exec_len - strlen(".manifest.sgx"), ".manifest.sgx")) {
+        } else if (strendswith(exec_uri, ".manifest.sgx")) {
             exec_uri[exec_len - strlen(".manifest.sgx")] = '\0';
         } else {
             SGX_DBG(DBG_E, "Invalid manifest file specified: %s\n", exec_uri);
```

dimakuv commented:
Jenkins, test this please

jiazhang0 commented:
@yamahata Could you help to speed up the completion of this RR? Thanks.

dimakuv commented:
Jenkins, retest Jenkins-SGX please (`vfork_and_exec` failed)

dimakuv commented:
Graphene-SGX fails on the bash example. It looks like your changes broke something in bash:
```sh
+ cd LibOS/shim/test/apps/bash
+ make SGX=1 all sgx-tokens
...
+ make SGX=1 regression

Building Bash Manifests

Run bash_test.sh:
./bash.manifest bash_test.sh 2
Sigstruct file not found ('sgx.sigfile' must be specified in manifest)
Makefile:20: recipe for target 'regression' failed
make: [regression] Error 234 (ignored)
```

Could you please debug it on your local machine?

jiazhang0 commented:
> Graphene-SGX fails on the bash example. It looks like your changes broke something in bash:
> 
> ```shell
> + cd LibOS/shim/test/apps/bash
> + make SGX=1 all sgx-tokens
> ...
> + make SGX=1 regression
> 
> Building Bash Manifests
> 
> Run bash_test.sh:
> ./bash.manifest bash_test.sh 2
> Sigstruct file not found ('sgx.sigfile' must be specified in manifest)
> Makefile:20: recipe for target 'regression' failed
> make: [regression] Error 234 (ignored)
> ```
> 
> Could you please debug it on your local machine?

Fixed and updated.

mkow commented:
Jenkins, retest this please

mkow commented:
Jenkins, test this please

mkow commented:
Jenkins, retest Jenkins-SGX please

mkow commented:
Jenkins, retest this please
